core/state/statedb.go

```diff
@@ -627,11 +627,36 @@ func (s *StateDB) createObject(addr common.Address) (newobj, prev *stateObject)
 	if prev == nil {
 		s.journal.append(createObjectChange{account: &addr})
 	} else {
+		// The original account should be marked as destructed and all cached
+		// account and storage data should be cleared as well. Note, it must
+		// be done here, otherwise the destruction event of original one will
+		// be lost.
 		_, prevdestruct := s.stateObjectsDestruct[prev.address]
 		if !prevdestruct {
 			s.stateObjectsDestruct[prev.address] = struct{}{}
 		}
-		s.journal.append(resetObjectChange{account: &addr, prev: prev, prevdestruct: prevdestruct})
+		var (
+			account []byte
+			storage map[common.Hash][]byte
+		)
+		// There may be some cached account/storage data already since IntermediateRoot
+		// will be called for each transaction before byzantium fork which will always
+		// cache the latest account/storage data.
+		if s.snapAccounts != nil {
```

Comment:
Perhaps small nitpick, instead of `if s.snapAccounts != nil {` and `if s.snapStorage != nil {`, use a single `if s.snap != nil` and have both bodies combined. That's how it's used elsewhere and IMHO is correct a sthe user won't assume that snapAccount can exist without snapStorage or the other way around.

Reply:
Yep, I will change it. But just want you know that I will maintain `accounts` and `storages` maps anyway no matter snap is available or not. Since I need to maintain the state diffs afterwards.